db/migrate/20220509104200_create_goods_nomenclature_export_function.rb

```diff
@@ -0,0 +1,120 @@
+            WHEN
+            RIGHT (gn.goods_nomenclature_item_id, 4) = '0000' THEN
+                6
+            WHEN
+            RIGHT (gn.goods_nomenclature_item_id, 2) = '00' THEN
+                8
+            ELSE
+                10
+            END AS significant_digits
+        FROM
+            goods_nomenclatures gn
+            JOIN goods_nomenclature_descriptions gnd ON gnd.goods_nomenclature_sid = gn.goods_nomenclature_sid
+            JOIN goods_nomenclature_description_periods gndp ON gndp.goods_nomenclature_description_period_sid = gnd.goods_nomenclature_description_period_sid
+            JOIN goods_nomenclature_indents gni ON gni.goods_nomenclature_sid = gn.goods_nomenclature_sid
+        WHERE (gn.validity_end_date IS NULL OR gn.validity_end_date >= coerced_actual_date)
+        AND gn.goods_nomenclature_item_id LIKE chapter_short_code
+        AND gndp.goods_nomenclature_description_period_sid IN (
+            SELECT
+                MAX(gndp2.goods_nomenclature_description_period_sid)
+            FROM
+                goods_nomenclature_description_periods gndp2
+            WHERE
+                gndp2.goods_nomenclature_sid = gnd.goods_nomenclature_sid
+                AND gndp2.validity_start_date <= coerced_actual_date)
+        AND gni.goods_nomenclature_indent_sid IN (
+            SELECT MAX(gni2.goods_nomenclature_indent_sid)
+            FROM goods_nomenclature_indents gni2
+            WHERE gni2.goods_nomenclature_sid = gn.goods_nomenclature_sid
+            AND gni2.validity_start_date <= coerced_actual_date
+        );
+
+            /* Index to speed up child node matching - need to perf test to see if any use */
+            CREATE INDEX t1_i_nomenclature ON tmp_nomenclature (goods_nomenclature_sid, goods_nomenclature_item_id);
+
+            /* Cursor loops through result set to identify if nodes are leaf and updates the flag if so */
+            DECLARE cur_nomenclature CURSOR FOR
+                SELECT
+                    *
+                FROM
+                    tmp_nomenclature;
+            BEGIN
+                FOR nom_record IN cur_nomenclature LOOP
+                    -- Raise Notice 'goods nomenclature item id %', nom_record.goods_nomenclature_item_id;
+                    /* Leaf nodes have to have pls of 80 and no children having the same nomenclature code */
+                    IF nom_record.producline_suffix = '80' THEN
+                        IF LENGTH(nom_record.node) = 10 OR NOT EXISTS (
+                            SELECT
+                                1
+                            FROM
+                                tmp_nomenclature
+                            WHERE
+                                goods_nomenclature_item_id LIKE CONCAT(nom_record.node, '%')
+                                    AND goods_nomenclature_item_id <> nom_record.goods_nomenclature_item_id)
+                                    AND number_indents <= nom_record.number_indents THEN
```

Comment:
So it turns out that this is actually wrong for a whole bunch of commodities even with this additional check on indents.

The grouping by node heuristic is actually not a precise way of knowing whether a goods nomenclature is part of the same tree. We have to go row by row to determine that. I think this is fine/is an improvement on what was there before but we should think about this at some stage